File: Algorithms/p373_Find_K_Pairs_with_Smallest_Sums/p373_Find_K_Pairs_with_Smallest_Sums.py
class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[List[int]]
        """
        if not nums1 or not nums2 or k==0:
            return []
        m = len(nums1)
        n = len(nums2)
        l = []
        if m<=n:
            for i in range(min(m,k)):
                heapq.heappush(l,(nums1[i]+nums2[0],i,0))
        else:
            for j in range(min(n,k)):
                heapq.heappush(l,(nums1[0]+nums2[j],0,j))
        res = []
        while k>0 and l:
            e = l[0]
            res.append([nums1[e[1]],nums2[e[2]]])
            if m<=n:
                if e[2]<n-1:
                    heapq.heappushpop(l,(nums1[e[1]]+nums2[e[2]+1],e[1],e[2]+1))
                else:
                    heapq.heappop(l)
            else:
                if e[1]<m-1:
                    heapq.heappushpop(l,(nums1[e[1]+1]+nums2[e[2]],e[1]+1,e[2]))
                else:
                    heapq.heappop(l)
            k -= 1
        return res

File: Algorithms/p373_Find_K_Pairs_with_Smallest_Sums/p373_Find_K_Pairs_with_Smallest_Sums.cpp
class Solution {
public:
    vector<pair<int, int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        if(nums1.empty()||nums2.empty()||k==0) return {};
        int m = nums1.size();
        int n = nums2.size();
        auto cmp = [&](pair<int,int>& a, pair<int,int>& b){ return nums1[a.first]+nums2[a.second]>nums1[b.first]+nums2[b.second]; };
        priority_queue<pair<int,int>,vector<pair<int,int>>,decltype(cmp)> heap(cmp);
        if(m<=n){
            for(int i=0;i<m&&i<k;++i){
                heap.emplace(i,0);
            }
        }else{
            for(int j=0;j<n&&j<k;++j){
                heap.emplace(0,j);
            }
        }
        vector<pair<int,int>> res;
        while(k>0&&!heap.empty()){
            auto e = heap.top();
            heap.pop();
            res.emplace_back(nums1[e.first],nums2[e.second]);
            if(m<=n){
                if(e.second<n-1) heap.emplace(e.first,e.second+1);
            }else{
                if(e.first<m-1) heap.emplace(e.first+1,e.second);
            }
            --k;
        }
        return res;
    }
};
